File: frontend/lib/types/prediction.ts
/**
 * TypeScript types for ML predictions
 */

export interface LSTMPredictionPoint {
  lap_number: number;
  predicted_lap_time: number;
  confidence_lower: number;
  confidence_upper: number;
}

export interface LSTMPredictionResponse {
  driver: string;
  compound: string;
  predictions: LSTMPredictionPoint[];
  model_rmse: number;
  training_laps_used: number;
  generated_at: string;
}

export interface FeatureImportance {
  feature: string;
  importance: number;
}

export interface RegressionPredictionResponse {
  predicted_lap_time: number;
  actual_lap_time: number | null;
  prediction_error: number | null;
  model_rmse: number;
  model_r2: number;
  feature_importances: FeatureImportance[];
  generated_at: string;
}

export interface PitStop {
  lap_number: number;
  compound_from: string;
  compound_to: string;
  expected_loss_seconds: number;
  confidence: number;
}

export interface StrategyOption {
  strategy_name: string;
  compounds: string[];
  pit_stops: PitStop[];
  expected_race_time: number;
  win_probability: number;
  percentile_10: number;
  percentile_90: number;
  risk_score: number;
}

export interface StrategyPredictionResponse {
  driver: string;
  current_lap: number;
  strategies: StrategyOption[];
  recommended_strategy: string;
  safety_car_probability: number;
  generated_at: string;
}

export interface ModelMetrics {
  model_type: string;
  rmse: number;
  mae: number;
  r2_score: number;
  training_samples: number;
  last_trained: string;
  accuracy_by_compound: Record<string, number>;
}

File: backend/app/schemas/prediction.py
"""
Pydantic schemas for ML predictions
"""

from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime


class LSTMPredictionRequest(BaseModel):
    """Request for LSTM lap time prediction"""
    race_id: int
    driver: str
    compound: str
    forecast_laps: int = Field(default=15, ge=1, le=30)


class LSTMPredictionPoint(BaseModel):
    """Single prediction point"""
    lap_number: int
    predicted_lap_time: float
    confidence_lower: float
    confidence_upper: float


class LSTMPredictionResponse(BaseModel):
    """Response for LSTM prediction"""
    driver: str
    compound: str
    predictions: List[LSTMPredictionPoint]
    model_rmse: float
    training_laps_used: int
    generated_at: datetime


class RegressionPredictionRequest(BaseModel):
    """Request for regression-based prediction"""
    race_id: int
    driver: str
    lap_number: int
    compound: str
    tyre_life: int
    track_temp: float
    air_temp: Optional[float] = None
    fuel_load: Optional[float] = None


class FeatureImportance(BaseModel):
    """Feature importance from regression model"""
    feature: str
    importance: float


class RegressionPredictionResponse(BaseModel):
    """Response for regression prediction"""
    predicted_lap_time: float
    actual_lap_time: Optional[float] = None
    prediction_error: Optional[float] = None
    model_rmse: float
    model_r2: float
    feature_importances: List[FeatureImportance]
    generated_at: datetime


class StrategyPredictionRequest(BaseModel):
    """Request for pit strategy prediction"""
    race_id: int
    driver: str
    current_lap: int = 1
    available_compounds: List[str] = Field(default=["SOFT", "MEDIUM", "HARD"])


class PitStop(BaseModel):
    """Predicted pit stop"""
    lap_number: int
    compound_from: str
    compound_to: str
    expected_loss_seconds: float
    confidence: float


class StrategyOption(BaseModel):
    """Single strategy option"""
    strategy_name: str
    compounds: List[str]
    pit_stops: List[PitStop]
    expected_race_time: float
    win_probability: float
    percentile_10: float  # Best case
    percentile_90: float  # Worst case
    risk_score: float  # 0-100, higher = riskier


class StrategyPredictionResponse(BaseModel):
    """Response for strategy prediction"""
    driver: str
    current_lap: int
    strategies: List[StrategyOption]
    recommended_strategy: str
    safety_car_probability: float
    generated_at: datetime


class ModelMetrics(BaseModel):
    """Overall model performance metrics"""
    model_type: str
    rmse: float
    mae: float
    r2_score: float
    training_samples: int
    last_trained: datetime
    accuracy_by_compound: Dict[str, float]
